modules/lib/test/Controller.test-d.ts

```diff
@@ -0,0 +1,187 @@
+import type {
+    Controller,
+    Request,
+    RequestBranch,
+    RequestMethod,
+    RequestMode,
+    RequestScheme,
+    Response,
+} from '../core/index';
+import {
+    expectAssignable,
+    // expectDeprecated,
+    // expectDocCommentIncludes,
+    // expectError,
+    // expectNever,
+	expectNotAssignable,
+    // expectNotDeprecated,
+    // expectNotType,
+    // expectType,
+    // printType,
+} from 'tsd';
+
+const log = {
+    info: (message?: string, ...optionalParams: string[]) => { /* no-op */ },
+};
+
+const myController = {
+    all: (request: Request) => {
+        log.info('all request:%s', JSON.stringify(request, null, 4));
+        return {
+            status: 200,
+        };
+    },
+    get: (request: Request) => {
+        log.info('get request:%s', JSON.stringify(request, null, 4));
+        return {
+            status: 200,
+        };
+    },
+    options: (request: Request) => {
+        log.info('options request:%s', JSON.stringify(request, null, 4));
+        return {
+            status: 200,
+        };
+    },
+    post: (request: Request) => {
+        log.info('post request:%s', JSON.stringify(request, null, 4));
+        return {
+            status: 200,
+        };
+    },
+};
+
+expectAssignable<Controller>(myController);
+
+type PageRequest = Omit<Request<{
+    // Only allow literal string
+    branch: RequestBranch
+    method: RequestMethod
+    mode: RequestMode
+    scheme: RequestScheme
+
+    // Make some optional properties required
+    repositoryId: string
+    webSocket: boolean
+}>,
+    // Omit/Disallow some optional properties
+    'contextPath' | 'validTicket'
+>;
+type PageRequestHandler = (request: PageRequest) => Response;
+
+interface PageController {
+    all?: PageRequestHandler
+    delete?: PageRequestHandler
+    get?: PageRequestHandler
+    head?: PageRequestHandler
+    options?: PageRequestHandler
+    post?: PageRequestHandler
+    put?: PageRequestHandler
+}
+
+const pageRequest /* : PageRequest */ = {
+    method: 'GET' as PageRequest['method'], // Avoid flattening down to string
```

Comment:
Doesn't the use of `as` go against the purpose of checking if this object satisfies a specific type?

Reply:
Yeah, but. I've added some comments to explain what's going on.

I'm specifically avoiding setting the type on the pageRequest object to PageRequest, so the assign test isn't just a dummy comparing PageRequest to PageRequest.

If you just copied the whole object into the assign test that would also work.

It's about how TypeScript "autogenerates" types for untyped things.

In normal code these problems shouldn't exist as the request object comes in from outside (doesn't even exist at compiletime), so the first thing we do is actually say that the shape of request is `AS Request`.